ddsrouter_core/include/ddsrouter_core/types/participant/ParticipantKind.hpp

```diff
@@ -32,96 +30,95 @@ namespace core {
-     */
-    enum
-    {
-        PARTICIPANT_KIND_INVALID,   //! Invalid Participant Kind
-        VOID,                       //! Void Participant Kind
-        ECHO,                       //! Echo Participant Kind
-        DUMMY,                      //! Dummy Participant Kind
-        SIMPLE_RTPS,                //! Simple RTPS Participant Kind
-        LOCAL_DISCOVERY_SERVER,     //! Discovery Server RTPS UDP Participant Kind
-        WAN,                        //! Discovery Server RTPS TCP Participant Kind
-    };
-
-    //! Default constructor that returns an Invalid Participant Kind
-    DDSROUTER_CORE_DllAPI ParticipantKind() noexcept;
-
-    //! Constructor by value of the ParticipantKind enum
-    DDSROUTER_CORE_DllAPI ParticipantKind(
-            ParticipantKindType value) noexcept;
-
-    /**
-     * @brief Whether this ParticipantKind is valid
-     *
-     * @return true if valid. False otherwise
-     */
-    DDSROUTER_CORE_DllAPI bool is_valid() const noexcept;
-
-    //! Convert this ParticipantKind to string using the << operator
-    DDSROUTER_CORE_DllAPI std::string to_string() const noexcept;
-
-    //! Return the enum value of this object
-    DDSROUTER_CORE_DllAPI ParticipantKindType operator ()() const noexcept;
-
-    //! Minor operator
-    DDSROUTER_CORE_DllAPI bool operator <(
-            const ParticipantKind& other) const noexcept;
-
-    //! Equal operator
-    DDSROUTER_CORE_DllAPI bool operator ==(
-            const ParticipantKind& other) const noexcept;
-
-    /**
-     * @brief Create a Participant Kind regarding the string argument in lower case
-     *
-     * @note Kind name is not case sensitive
-     *
-     * It compares the argument \c kind in lower case with any of the existing kind names, and in case it
-     * matches any of them, return the ParticipantKind associated with that name.
-     * It will return \c INVALID in case no existing ParticipantKind name matches the argument.
-     *
-     * @param [in] kind : string with the name of the kind to build
-     * @return ParticipantKind value, \c INVALID if \c kind does not refer to any existing kind
-     */
-    DDSROUTER_CORE_DllAPI static ParticipantKind participant_kind_from_name(
-            std::string kind) noexcept;
-
-    DDSROUTER_CORE_DllAPI static std::vector<ParticipantKind> all_valid_participant_kinds() noexcept;
-
-protected:
-
-    //! Value that links with the enumeration ParticipantKind of this object
-    ParticipantKindType value_;
-
-    //! Map that link a \c ParticipantKind with its name to be created or deserialized
-    static const std::map<ParticipantKindType, std::vector<std::string>> participant_kind_with_aliases_;
-
-    // Allow operator << to use \c participant_kind_with_aliases_
-    DDSROUTER_CORE_DllAPI friend std::ostream& operator <<(
-            std::ostream& os,
-            const ParticipantKind& kind);
+constexpr std::array<ParticipantKind, ParticipantKindCount> AllParticipantKinds = {
```

Comment:
I get the idea behind this implementation, but I am not into it. For me it does not seems clean neither easy to read or change.
Please @rsanchez15 and @juanlofer-eprosima check it and give your opinions.

In my opinion, we could work on some how of creating an enumeration class or macro that will create all of this generally, so we do not need these magic numbers and const variables all over when we need to create an enumeration related with a string.
(Somehow Python do it)

Reply:
This is an old topic in the C++ communiity. The lack of a reflection mechanism in C++ (see proposal https://en.cppreference.com/w/cpp/keyword/reflexpr) makes impossible to do that without macros, but macros are discouraged in modern C++ for obvious reasons. I already have a macro implementation that does this, you write `ENUM_DEFINITION(enumName, enumVal1, enumVal2, ...)` and these kind of things are automatically defined, so it would be a quick addition. Let's discuss this in a separate task.